src/main/java/org/dataone/hashstore/filehashstore/FileHashStore.java

```diff
@@ -981,249 +683,288 @@ public InputStream retrieveMetadata(String pid, String formatId)
-            );
-        } catch (IOException ioe) {
-            String errMsg =
-                "FileHashStore.retrieveMetadata - Unexpected error when creating InputStream"
-                    + " for pid: " + pid + " with formatId: " + DEFAULT_METADATA_NAMESPACE
-                    + ". IOException: " + ioe.getMessage();
-            logFileHashStore.error(errMsg);
-            throw new IOException(errMsg);
-        }
+            // Before we begin deletion process, we look for the `cid` by calling
+            // `findObject` which will throw custom exceptions if there is an issue with
+            // the reference files, which help us determine the path to proceed with.
+            try {
+                Map<String, String> objInfoMap = findObject(pid);
+                String cid = objInfoMap.get("cid");
 
-        return metadataCidInputStream;
-    }
+                // If no exceptions are thrown, we proceed to synchronization based on the `cid`
+                synchronizeReferencedLockedCids(cid);
 
-    @Override
-    public void deleteObject(String cid, boolean deleteCid) throws IllegalArgumentException,
-        FileNotFoundException, IOException, NoSuchAlgorithmException {
-        logFileHashStore.debug(
-            "FileHashStore.deleteObject - Called to delete object with content identifeir: " + cid
-        );
-        if (deleteCid) {
-            // Validate input parameters
-            FileHashStoreUtility.ensureNotNull(cid, "cid", "deleteObject");
-            FileHashStoreUtility.checkForEmptyString(cid, "cid", "deleteObject");
+                try {
+                    // Proceed with comprehensive deletion - cid exists, nothing out of place
+                    Path absCidRefsPath = getHashStoreRefsPath(cid, HashStoreIdTypes.cid.getName());
+                    Path absPidRefsPath = getHashStoreRefsPath(pid, HashStoreIdTypes.pid.getName());
+
+                    // Begin deletion process
+                    updateRefsFile(pid, absCidRefsPath, "remove");
+                    if (Files.size(absCidRefsPath) == 0) {
+                        Path objRealPath = getHashStoreDataObjectPath(pid);
+                        deleteList.add(FileHashStoreUtility.renamePathForDeletion(objRealPath));
+                        deleteList.add(FileHashStoreUtility.renamePathForDeletion(absCidRefsPath));
+                    } else {
+                        String warnMsg = "cid referenced by pid: " + pid
+                            + " is not empty (refs exist for cid). Skipping object deletion.";
+                        logFileHashStore.warn(warnMsg);
+                    }
+                    deleteList.add(FileHashStoreUtility.renamePathForDeletion(absPidRefsPath));
+                    // Delete all related/relevant items with the least amount of delay
+                    FileHashStoreUtility.deleteListItems(deleteList);
+                    deleteMetadata(pid);
```

Comment:
If the method `deleteMetadata(pid)` fails, do you have a feature to recover the deleted files on line 738?

Reply:
I have added a feature to recover deleted files and revert the process if an unexpected issue occurs. Thank you for suggesting this change!